File: Solutions/14. Longest Common Prefix/CPP/Longest_Common_Prefix.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution{
public:
    string longestCommonPrefix(vector<string> &strs){
        string result = "";
        int counter = 0;
        while(counter < strs[0].size()){ //return if the first string is already empty
            for(int i = 1; i < strs.size(); i++){
                if(counter >= strs[i].size() or strs[i][counter] != strs[0][counter]){
                    return result;
                }
            }
            result.push_back(strs[0][counter]);
            counter = counter + 1;
        }
        return result;
    }
};

int main(){
    Solution* S = new Solution();

    //input
    vector<string> strs{"a", "abc", "apple"};

    cout << S->longestCommonPrefix(strs) << endl;
}

File: Solutions/14. Longest Common Prefix/Ruby/Longest_Common_Prefix.rb
def longest_common_prefix(strs)
    length = strs.map{ |str| str.length}.min

    prefix = ""
    0.upto(length - 1) do |i|
        char = strs[0][i] # use strs[0] as the template
        strs.each do |str|
            if str[i] != char
                return prefix
            end
        end

        prefix += char
    end

    prefix
end

strs = ["a", "abc", "apple"]

puts longest_common_prefix(strs)